pkg/codegen/internal/test/sdk_driver.go

```diff
@@ -4,23 +4,73 @@ import (
 	"os"
 	"os/exec"
 	"path/filepath"
+	"strings"
 	"testing"
 
-	"github.com/pulumi/pulumi/pkg/v3/codegen"
+	//"github.com/pulumi/pulumi/pkg/v3/codegen"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
-type ThenFunc func(t *testing.T, testDir string)
+type SdkTest struct {
+	// Short name of the directory, relative to `TestDir`.
+	Directory string
 
-type sdkTest struct {
-	Directory   string
 	Description string
-	Skip        codegen.StringSet
-	Then        map[string]ThenFunc
+
+	// Decides which programming languages to test the code
+	// generation for; defaults to all supported languages.
+	IncludeLanguage func(lang string) bool
+
+	// Hooks that run additional validations after the code has
+	// been generated. Hooks are usually language-specific and are
+	// included in this map by language.
+	HooksByLanguage map[string][]SdkTestHook
+}
+
+type SdkTestHook struct {
+	// Name such as "compile", "test".
+	Name string
+
+	// Checks to run.
+	RunHook func(env *SdkTestEnv)
+}
+
+type SdkTestEnv struct {
+	T *testing.T
+
+	// The directory with the generated code, typically `$TestDir/$Language`.
+	CodeDir string
+}
+
+func (env *SdkTestEnv) NewCommandInDir(dir string, name string, arg ...string) *exec.Cmd {
+	d := filepath.Join(env.CodeDir, dir)
+	env.T.Logf("cd %s && %s %s", d, name, strings.Join(arg, " "))
```

Comment:
Helpers for invoking sub-commands. I really like having the command printed out so I can  just copy and try it in my shell if I want to retry that particular step.